builtin/mainmenu/dlg_file_browser.lua

```diff
@@ -0,0 +1,184 @@
+-- Minetest
+-- Copyright (C) 2020 Jean-Patrick Guerrero <jeanpatrick.guerrero@gmail.com>
+-- Copyright (C) 2020 EvidenceBKidscode
+
+-- This program is free software; you can redistribute it and/or modify
+-- it under the terms of the GNU Lesser General Public License as published by
+-- the Free Software Foundation; either version 2.1 of the License, or
+-- (at your option) any later version.
+--
+-- This program is distributed in the hope that it will be useful,
+-- but WITHOUT ANY WARRANTY; without even the implied warranty of
+-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+-- GNU Lesser General Public License for more details.
+--
+-- You should have received a copy of the GNU Lesser General Public License along
+-- with this program; if not, write to the Free Software Foundation, Inc.,
+-- 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+local ESC = core.formspec_escape
+local PATH = os.getenv("HOME") or os.getenv("HOMEPATH") or core.get_worldpath()
+local tabdata = {}
+
+local function filesize(name)
+	local f = io.open(name)
+	if not f then return end
+	local size = f:seek("end")
+	f:close()
+
+	return size
+end
+
+local function get_dirs()
+	local dirs
+
+	if PATH == "\\" then -- Windows root dir
+		dirs = minetest.get_dir_list(os.getenv("HOMEDRIVE"))
+	else
+		dirs = minetest.get_dir_list(PATH)
+	end
+
+	local new = {}
+	for _, f in ipairs(dirs) do
+		if f:sub(1,1) ~= "." then
+			local is_file = f:match("^.+(%..+)$")
+
+			if not is_file then
+				new[#new + 1] = f
+			end
+		end
+	end
+
+	table.sort(new)
+	return new
+end
+
+local function make_fs()
+	local dirs = get_dirs()
+	local _path, value = "", ""
+
+	for dirname in string.gmatch(PATH, "([^" .. DIR_DELIM .. "]+)") do
+		value = value .. DIR_DELIM:gsub("\\", "\\\\") .. (dirname:gsub("_", "_u"):gsub(" ", "_s"))
+		_path = ("%s %s <action name=%s>%s</action>")
+			:format(_path, DIR_DELIM:gsub("\\", "\\\\\\\\"), value, dirname)
+	end
+
+	local fs = "size[10,7]" ..
+		"real_coordinates[true]" ..
+		"image_button[0.2,0.15;0.5,0.5;" ..
+			ESC(defaulttexturedir .. "arrow_up.png") .. ";updir;]" ..
```

Comment:
This for instance would accordingly be `texture("arrow_up")`. This entire formspec however should IMO be a string to be formatted using `string.format`.